### ISSUE TITLE
strptime(.., '%c') fails to parse output of strftime('%c', ..) in some locales

### ISSUE THREAD

**Issue by abalkin** [python/cpython#53203]
BPO | [8957](https://bugs.python.org/issue8957)
--- | :---
Nosy | @abalkin, @vstinner, @ezio-melotti
Dependencies | <li>bpo-8915: Use locale.nl_langinfo in _strptime</li>
Files | <li>[strptime-locale-bug.c](https://bugs.python.org/file17601/strptime-locale-bug.c "Uploaded as application/octet-stream at 2010-06-09.20:24:38 by @abalkin"): Working C code</li><li>[strptime-locale-bug.py](https://bugs.python.org/file17602/strptime-locale-bug.py "Uploaded as text/plain at 2010-06-09.20:25:05 by @abalkin"): Failing python code</li><li>[cfmt.py](https://bugs.python.org/file20358/cfmt.py "Uploaded as text/plain at 2011-01-12.00:08:40 by @abalkin")</li><li>[issue8957.py3k.1.patch](https://bugs.python.org/file20412/issue8957.py3k.1.patch "Uploaded as text/plain at 2011-01-15.07:20:49 by eli.bendersky")</li>

<sup>*Note: these values reflect the state of the issue at the time it was migrated and might not reflect the current state.*</sup>

<details><summary>Show more details</summary><p>

GitHub fields:
```python
assignee = None
closed_at = None
created_at = <Date 2010-06-09.20:24:38.889>
labels = ['extension-modules', 'type-bug', '3.7']
title = "strptime(.., '%c') fails to parse output of strftime('%c', ..) in some locales"
updated_at = <Date 2016-09-26.21:29:29.654>
user = 'https://github.com/abalkin'
```

bugs.python.org fields:
```python
activity = <Date 2016-09-26.21:29:29.654>
actor = 'belopolsky'
assignee = 'none'
closed = False
closed_date = None
closer = None
components = ['Extension Modules']
creation = <Date 2010-06-09.20:24:38.889>
creator = 'belopolsky'
dependencies = ['8915']
files = ['17601', '17602', '20358', '20412']
hgrepos = []
issue_num = 8957
keywords = ['patch']
message_count = 18.0
messages = ['107413', '125966', '125968', '126043', '126045', '126055', '126057', '126059', '126084', '126235', '126313', '126316', '126339', '126340', '126345', '126347', '126358', '221924']
nosy_count = 4.0
nosy_names = ['belopolsky', 'vstinner', 'ezio.melotti', 'rpetrov']
pr_nums = []
priority = 'normal'
resolution = None
stage = 'needs patch'
status = 'open'
superseder = None
type = 'behavior'
url = 'https://bugs.python.org/issue8957'
versions = ['Python 3.7']
```

</p></details>

### Linked PRs
* gh-124946
* gh-125090
* gh-125091
* gh-125093
* gh-125369
* gh-125370
* gh-125406
* gh-125454
* gh-125469
* gh-125483
* gh-125785
* gh-125806
* gh-125807
* gh-135971
* gh-136019
* gh-136020

**Comment by serhiy-storchaka** [comment]
I agree that using the OS `strftime()` or at least `nl_langinfo()` is more preferable, but these issues are already 14 years old. There may be issues:

* We cannot guarantee that `strftime()` exists and works on all platforms, so we still need a fallback.
* Due to a bug in `nl_langinfo()` (and a flaw of the `locale` module), it does not work on some locales with non-UTF-8 default encoding. Even if it works, it can returns not currently supported format codes like `%OY` or `%R`. So we still need a pseudo-AI as a fallback.

So I took the Eli's patch and added numerous tests for `strftime()` on different locales (they were written on another occasion).

**Comment by serhiy-storchaka** [comment]
I tested with many locales and found several cases in which this patch did not work:

* In Basque, the month name is only used in %x. %c uses month number. So the month format should be searched separately for different formats.
* In Arabic, the day of the week name and the AM/PM indicator can be found as the part of the month name. This was a regression. Thus, the month name should be replaced before most of other items. But this breaks Japanese, because the full month name "3月" is a substring of "03月" which was created as "%m月" in %c. Thus, we should check the month name is found at the same position for all months. But this breaks Morisyen, because the month name "me" is a substring of the day of the week abbreviated name "mer". Thus, we should check all positions for all months.

There are similar bugs related to ambiguity of the day of the week name. For example, in Kashubian, the month name matches the day of the week name for the sample datetime. In Yoruba, an abbreviated day names are used with suffix "ọjọ ́", and this matches the full day name, but not all full day name have such suffix.

Many locales fail due to space-padded numbers or non-ASCII digits.

I am not sure what of these bugs I'll fix here and what will leave for other issues.

**Comment by serhiy-storchaka** [comment]
For now, this PR fixes `strptime()` for `%c` and `%x` formats in multiple locales for the following languages: Arabic, Bislama, Chuvash, Estonian, French, Irish, Gurajati, Manx Gaelic, Hebrew, Hindi, Chhattisgarhi, Haitian Kreyol, Japanese, Kannada, Korean, Marathi, Malay, Norwegian, Nynorsk, Punjabi, Rajasthani, Tok Pisin, Yue Chinese, Yau/Nungon and Chinese.

**Comment by nineteendo** [comment]
Is this fixed now?

**Comment by serhiy-storchaka** [comment]
Not yet. Several other locales can be fixed using other approach. I am working on this.

**Comment by serhiy-storchaka** [comment]
#125406 fixes most of other locales that use non-ASCII digits.

**Comment by kulikjak** [comment]
Hi, testing all the merged changes on Solaris, I encountered two issues that I wanted to report:

1. `test_date_time_locale2` started failing now that year 1800 was added:

```
  File "/..../Lib/_strptime.py", line 436, in _strptime
    raise ValueError("time data %r does not match format %r" %
ValueError: time data 'January  1, 1800 at 12:00:00 AM LMT' does not match format '%c'
```

Based on my digging, I believe that the issue is in the LMT; I dumped the `processed_format` used for the check, and it doesn't include LMT (it ends with: `(?P<Z>cest|gmt|utc|cet)`)

I have no idea where that LMT comes from or whether it is Solaris specific though. It's there for old years:
```python
import locale
import time
locale.setlocale(locale.LC_TIME, "de_DE")
print(time.strftime("%c", (1900, 1, 1, 0, 0, 0, 0, 1, 0)))
print(time.strftime("%c", (1800, 1, 1, 0, 0, 0, 0, 1, 0)))
```
```
 1. Januar 1900 um 00:00:00 Uhr CET
 1. Januar 1800 um 00:00:00 Uhr LMT
```

2. `%x` on Solaris returns only the last two digits of the year for many locales, meaning that `test_date_locale2` often fails as it's not possible to distinguish for example 1900 from 2000.

I don't think that the output of %x is standardized (Windows format is also different, although the year seems always 4 digit)? So it's possible that this might be an issue for other platforms as well.

**Comment by serhiy-storchaka** [comment]
Thank you @kulikjak.

LMT is a standard thing. I think the result of the test depends on the place where they are run. I do not know what to do with LMT, so I am going to just ignore failures if the result contains LMT. This may be a part of a larger issue.

> `%x` on Solaris returns only the last two digits of the year for many locales

What locales? We can skip this test on Solaris, but it would be better to keep it running for locales in which it works.

**Comment by kulikjak** [comment]
> ... so I am going to just ignore failures if the result contains LMT. This may be a part of a larger issue.

That makes sense. Thank you.

> What locales? We can skip this test on Solaris, but it would be better to keep it running for locales in which it works.

From those being tested, _en_US_, _de_DE_ and _ar_AE_ do print only two digits with `%x`.

**Comment by terryjreedy** [comment]
@serhiy-storchaka Can we close this?  It looks to me that any further issues should be new issues.

**Comment by serhiy-storchaka** [comment]
I was going to create one or two more PRs before closing this PR.

**Comment by StanFromIreland** [comment]
Did you manage to open them, can this be closed now?

**Comment by serhiy-storchaka** [comment]
#135971 is perhaps the last PR for this issue. It adds support of non-decimal alt digits on locale lzh_TW, fixes bug in detecting the position of week day name on locales byn_ER and wal_ET, fixes support of single-digit hour on locales ar_SA and bg_BG.

The  code is prepared for using nl_langinfo(). The rest of changes belongs to #53161.
